README.rst
..
    This file is part of BDC-Auth-Client.
    Copyright (C) 2022 INPE.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.


========================================================================
A Client Package for Authentication and Authorization Based on OAuth 2.0
========================================================================


.. image:: https://img.shields.io/badge/License-GPLv3-blue.svg
        :target: https://github.com//brazil-data-cube/bdc-auth-client/blob/master/LICENSE
        :alt: Software License


.. image:: https://readthedocs.org/projects/bdc-auth-client/badge/?version=latest
        :target: https://bdc-auth-client.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status


.. image:: https://img.shields.io/badge/lifecycle-stable-green.svg
        :target: https://www.tidyverse.org/lifecycle/#stable
        :alt: Software Life Cycle


.. image:: https://img.shields.io/github/tag/brazil-data-cube/bdc-auth-client.svg
        :target: https://github.com/brazil-data-cube/bdc-auth-client/releases
        :alt: Release


.. image:: https://img.shields.io/discord/689541907621085198?logo=discord&logoColor=ffffff&color=7389D8
        :target: https://discord.com/channels/689541907621085198#
        :alt: Join us at Discord


About
=====


BDC-Auth-Client is a client package for authentication and authorization based on OAuth 2.0.


Installation
============


See `INSTALL.rst <./INSTALL.rst>`_.


Usage
=====


See `USAGE.rst <./USAGE.rst>`_.


Developer Documentation
=======================


See https://bdc-auth-client.readthedocs.io/en/latest.


License
=======


.. admonition::
    Copyright (C) 2022 INPE.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.


CHANGES.rst
..
    This file is part of BDC-Auth-Client.
    Copyright (C) 2022 INPE.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.


Changes
=======


Version 0.4.3 (2022-11-23)
--------------------------

- Expand all keys from token introspection in request context (`#36 <https://github.com/brazil-data-cube/bdc-auth-client/issues/36>`_)


Version 0.4.2 (2022-09-22)
--------------------------

- Change LICENSE to GPL v3


Version 0.4.1 (2022-09-08)
--------------------------

- Fix exception when OAuth2 servers delivery a expired token (Authlib 1.0+) (`#31 <https://github.com/brazil-data-cube/bdc-auth-client/issues/31>`_)
- Update license headers date


Version 0.4.0 (2022-02-14)
--------------------------


- Improve log handler for expired tokens and generic errors `#16 <https://github.com/brazil-data-cube/bdc-auth-client/issues/16>`_.
- Improve documentation and usage `#9 <https://github.com/brazil-data-cube/bdc-auth-client/issues/9>`_.


Version 0.2.3 (2021-08-12)
--------------------------


- Add support for OR operations `#25 <https://github.com/brazil-data-cube/bdc-auth-client/issues/25>`_.

- Remove travis CI `#24 <https://github.com/brazil-data-cube/bdc-auth-client/issues/24>`_.

- Add drone support `#18 <https://github.com/brazil-data-cube/bdc-auth-client/issues/18>`_.

Version 0.2.2 (2021-07-12)
--------------------------


- Retrieve user id in decorators `#20 <https://github.com/brazil-data-cube/bdc-auth-client/issues/20>`_.


Version 0.2.1 (2022-08-26)
--------------------------


- Bug Fix in decorator oauth2: See `commit aa436 <https://github.com/brazil-data-cube/bdc-auth-client/commit/aa43602d25063678e69ba6ff6bd84653a7b20e2b>`_.



Version 0.2.0 (2022-08-20)
--------------------------


- First experimental version.
- Based on `BDC-Auth <https://github.com/brazil-data-cube/bdc-auth>`_.
- Based on OAuth 2.0 Authorization Framework - `RFC6749 <https://tools.ietf.org/html/rfc6749>`_.
- Based on `Authlib <https://authlib.org/>`_.
- Documentation system based on Sphinx.
- Package support through Setuptools.
- Installation instructions.
- Source code versioning based on `Semantic Versioning 2.0.0 <https://semver.org/>`_.
- License: `MIT <https://github.com/brazil-data-cube/bdc-auth-client/blob/master/LICENSE>`_.


run-tests.sh
#!/usr/bin/env bash
#
# This file is part of BDC-Auth-Client.
# Copyright (C) 2022 INPE.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
#

pydocstyle bdc_auth_client setup.py && \
isort bdc_auth_client setup.py --check-only --diff && \
check-manifest --ignore ".drone.yml,.readthedocs.yml" && \
sphinx-build -qnW --color -b doctest docs/sphinx/ docs/sphinx/_build/doctest #&& \
pytest

tests/test_bdc_auth_client.py
#
# This file is part of BDC-Auth-Client.
# Copyright (C) 2022 INPE.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
#

"""Unit-test for BDC-Auth-Client."""

from flask import Flask
from werkzeug.exceptions import HTTPException, InternalServerError

from bdc_auth_client.decorators import oauth2


TOKEN = 'Fake'
ROUTE = '/'
HOST = 'https://localhost'


def setup_app(throw_exception=True, required=True):
    """Create a Flask application with a minimal routes and testing purposes."""
    app = Flask(__name__)
    app.config['BDC_AUTH_ACCESS_TOKEN_URL'] = HOST
    app.config['BDC_AUTH_CLIENT_ID'] = ''
    app.config['BDC_AUTH_CLIENT_SECRET'] = ''

    @app.route(ROUTE)
    @oauth2(roles=['admin'], throw_exception=throw_exception, required=required)
    def index(**kwargs):
        """Return the protected index route."""
        return dict(msg='OK')

    @app.errorhandler(Exception)
    def handle_exception(e):
        """Handle exceptions."""
        if isinstance(e, HTTPException):
            return {'code': e.code, 'description': e.description}, e.code

        app.logger.exception(e)

        return {'code': InternalServerError.code,
                'description': InternalServerError.description}, InternalServerError.code

    return app


def assert_json(response, code, msg=None):
    """Define a assert utility for JSON API responses."""
    assert response.content_type == 'application/json'
    assert response.status_code == code
    if msg:
        assert response.json['code'] == code
        assert response.json['description'] == msg


def get_app_client(**options):
    """Wrap the flask application and return a test client."""
    app = setup_app(**options)

    return app.test_client()


def test_decorator_token_expired(requests_mock):
    """Test the token expiration and message."""
    client = get_app_client()

    # Token Expired
    requests_mock.post(HOST, json=dict(status=False),
                       status_code=200,
                       headers={'content-type': 'application/json'})
    resp = client.get(ROUTE, query_string=dict(access_token=TOKEN))
    assert_json(resp, 401, 'Token expired.')


def test_decorator_forbidden(requests_mock):
    """Test the default behavior for decorator (User without right permission)."""
    client = get_app_client()

    requests_mock.post(HOST, json=dict(status=True, sub=dict(roles=['read'])),
                       status_code=200,
                       headers={'content-type': 'application/json'})
    resp = client.get(ROUTE, query_string=dict(access_token=TOKEN))
    assert_json(resp, 403, 'You don\'t have permission to access this resource.')


def test_decorator_missing_token(requests_mock):
    """Test when the user does not provide any token."""
    client = get_app_client()

    # Missing TOKEN (Required)
    requests_mock.post(HOST, json=dict(status=True, sub=dict(roles=['admin'])),
                       status_code=200,
                       headers={'content-type': 'application/json'})
    resp = client.get(ROUTE)
    assert_json(resp, 401, 'Missing access_token parameter.')


def test_token_optional(requests_mock):
    """Test when access token is optional."""
    client_non_required = get_app_client(required=False)
    # Missing TOKEN (non-required)
    requests_mock.post(HOST, json=dict(status=True, sub=dict(roles=['admin'])),
                       status_code=200,
                       headers={'content-type': 'application/json'})
    resp = client_non_required.get(ROUTE)
    assert_json(resp, 200)
    assert resp.json['msg'] == 'OK'


def test_remote_server_internal_error(requests_mock):
    """Test when occurs a error using decorator."""
    client = get_app_client()

    expected_code = 500
    expected_msg = InternalServerError.description

    # Token OK
    requests_mock.post(HOST, json=dict(),
                       status_code=expected_code,
                       headers={'content-type': 'application/json'})
    resp = client.get(ROUTE, query_string=dict(access_token=TOKEN))

    assert_json(resp, expected_code, expected_msg)


def test_decorator(requests_mock):
    """Test well-known way to validate a token."""
    client = get_app_client()

    # Token OK
    requests_mock.post(HOST, json=dict(status=True, sub=dict(roles=['admin'])),
                       status_code=200,
                       headers={'content-type': 'application/json'})
    resp = client.get(ROUTE, query_string=dict(access_token=TOKEN))
    assert_json(resp, 200)
    assert resp.json['msg'] == 'OK'

    # Test again to use cache
    resp = client.get(ROUTE, query_string=dict(access_token=TOKEN))
    assert_json(resp, 200)


bdc_auth_client/__init__.py
#
# This file is part of BDC-Auth-Client.
# Copyright (C) 2022 INPE.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
#

"""User authentication and authorization system for BDC."""

from .decorators import oauth2
from .version import __version__

__all__ = (
    'oauth2',
    '__version__',
)


bdc_auth_client/decorators.py
#
# This file is part of BDC-Auth-Client.
# Copyright (C) 2022 INPE.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
#

"""Decorators used to integrate with BDC-Auth Provider."""

from functools import wraps

from authlib.integrations.requests_client import OAuth2Session
from cacheout.cache import Cache
from flask import abort, current_app, request
from werkzeug.exceptions import HTTPException

# Define a InMemory cache for development purpose
# Used to prevent `fetch_token` all the time.
token_cache = Cache(maxsize=512, ttl=600)

HTTP_403_MSG = 'You don\'t have permission to access this resource.'
"""Define a generic message related Fordibben Access (403)."""


def oauth2(roles=None, required=True, throw_exception=True):
    """Decorate a Flask route to connect with BDC-Auth Provider.

    You can specify user roles required to access a resource.

    Make sure to set the following variables to the app:

    - ``BDC_AUTH_CLIENT_ID``: Application Client Id
    - ``BDC_AUTH_CLIENT_SECRET``: Application Client Secret
    - ``BDC_AUTH_ACCESS_TOKEN_URL``: URL to BDC-Auth Provider Token

    Example:
        >>> from bdc_auth_client.decorators import oauth2
        >>>
        >>> @app.route('/')  # doctest: +SKIP
        ... @oauth2(roles=['admin'])
        ... def protected_route(roles=None):
        ...     return dict(status=200)
    """
    def _oauth2(func):
        @wraps(func)
        def wrapped(*args, **kwargs):
            access_token = request.headers['x-api-key'] if request.headers.get('x-api-key') else \
                            request.args.get('access_token')
            if not access_token:
                if required:
                    abort(401, 'Missing access_token parameter.')
                return func(*args, **kwargs)

            if token_cache.has(access_token):
                res = token_cache.get(access_token)
                kwargs.update(dict(roles=res['sub']['roles'] or []))
                kwargs.update(dict(access_token=access_token))
                kwargs.update(dict(user_id=res.get('user_id', None)))

            else:
                session = OAuth2Session(
                    client_id=current_app.config['BDC_AUTH_CLIENT_ID'],
                    client_secret=current_app.config['BDC_AUTH_CLIENT_SECRET'],
                    token_endpoint_auth_method='client_secret_basic',
                )
                try:
                    # Set policy to access the `BDC-Auth`
                    # A policy is associated with Roles.
                    policy = None
                    policies = []

                    if roles:
                        for role in roles:
                            if not isinstance(role, (list, tuple)):
                                policies.append(role)
                            else:
                                for r in role:
                                    policies.append(r)

                        policy = ','.join(policies)

                    res = session.fetch_token(
                        current_app.config['BDC_AUTH_ACCESS_TOKEN_URL'], grant_type='introspect',
                        token=access_token, policy=policy
                    )

                    # Token Expired
                    if 'status' in res and not res['status']:
                        abort(401, 'Token expired.')

                    active = res.get('active')
                    if 'code' in res or (active is not None and not active):
                        abort(403, HTTP_403_MSG)

                    user_roles = res['sub'].get('roles', [])
                    kwargs.update(dict(roles=user_roles))
                    kwargs.update(dict(access_token=access_token))
                    kwargs.update(dict(user_id=res.get('user_id', None)))
                    for key, value in res['sub'].items():
                        if key != 'roles':
                            kwargs.update({key: value})

                    if roles:
                        for role in roles:
                            if isinstance(role, (list, tuple)):
                                if not set(role).intersection(set(user_roles)):
                                    abort(403, HTTP_403_MSG)
                                continue

                            if role not in user_roles:
                                return abort(403, HTTP_403_MSG)

                    token_cache.add(access_token, res, ttl=60)
                except HTTPException:
                    if throw_exception:
                        raise
                except Exception as e:
                    if throw_exception:
                        abort(500)
            return func(*args, **kwargs)
        return wrapped
    return _oauth2
